torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
 del method
 del func
 
+try:
+    import z3
+
+    Z3Expr = Union[z3.ArithRef, z3.BoolRef]
+
+    def _boolref(expr: Any) -> z3.BoolRef:
+        return cast(z3.BoolRef, expr)
+
+    # Wrapper of Z3 expressions.
+    #
+    # Some operations on Z3 depend on an assertion so that Z3 can reason
+    # about it correctly. e.g. division depends on the denominator being
+    # different than 0.
+    #
+    # This class helps propagate such an assertion to the caller.
+    @dataclass
+    class Z3Node:
+        # Actual Z3 expression.
+        expr: Z3Expr
+        # Assertions for this expression.
+        assertion: Optional[z3.BoolRef] = None
+
+        @property
+        def arith(self) -> z3.ArithRef:
+            assert isinstance(self.expr, z3.ArithRef)
+            return self.expr
+
+        @property
+        def bool(self) -> z3.BoolRef:
+            assert isinstance(self.expr, z3.BoolRef)
+            return self.expr
+
+    # Implementation of Python semantics as Z3 expressions.
+    #
+    # Z3 Real-Int theory has operators with semantics that differ that of
+    # Python. Therefore, in order to get it right, we need to implement
+    # the (Python) semantics we are relying on in Z3.
+    class Z3Ops:
+        # The 2 functions below are used for conditionally casting between
+        # integer and reals.
+        #
+        # Returns a real expression from 'x'.
+        @staticmethod
+        def to_real(x: z3.ArithRef) -> Z3Node:
+            return Z3Node(x if x.is_real() else z3.ToReal(x))
+
+        # Returns an integer expression from 'x'.
+        @staticmethod
+        def to_int(x: z3.ArithRef) -> Z3Node:
+            return Z3Node(x if x.is_int() else z3.ToInt(x))
+
+        # Implements Python division semantics.
+        # This is needed because Z3 integer division does not have the
+        # same semantics as Python true division.
+        @staticmethod
+        def div(numerator: z3.ArithRef, denominator: z3.ArithRef) -> Z3Node:
+            denominator = Z3Ops.to_real(denominator).arith
+            return Z3Node(
+                expr=Z3Ops.to_real(numerator).arith / denominator,
+                assertion=_boolref(denominator != 0),
+            )
+
+        @staticmethod
+        def floor(number: z3.ArithRef) -> Z3Node:
+            return Z3Node(Z3Ops.to_int(number).expr)
```

Comment:
It is interesting you just drop the assertion here. In this case I suppose you know assertion is None because a division didn't happen, but it's kind of non-uniform (and if you ever modified the assertion to hold more interesting stuff, this could lead to bugs). I'm still more keen on separately guarding on not zero IMO.

Reply:
What if, instead, `Z3Ops` held a reference to a `TranslationValidator` object? So that we would access it like: `Z3Ops(validator).<operator>(<args>)`?

Reply:
sure